### PR TITLE
ARROW-6736: [Rust] [DataFusion] Evaluate the input to the aggregate expression just once per batch

### PR DESCRIPTION
**Pull request by andygrove**
The current implementation of aggregate expressions in the new physical plan had a flaw where the input to the aggregate expression was repeatedly being evaluated (once per row instead of once per batch). This PR fixes this.

### CODE REVIEW

**Comment by andygrove** [comment]
@sinistersnare Could you review this? This is blocking your PRs being merged. Thanks!

**Comment by andygrove** [comment]
@alippai Could you review this? This is blocking your PRs being merged. Thanks!

**Comment by andygrove** [comment]
Also could could I get a review from a committer please @paddyhoran @sunchao @nevi-me

**Comment by alippai** [comment]
@andygrove looks good! Nit: was adding `input` as a parameter for `create_accumulator()` considered? It would keep the iterations&maps easier to read later (the downside it's adding extra state to the accumulator)

**Comment by sinistersnare** [comment]
I also think it looks good for the sum/count use-case, and also for new cases.

**Comment by andygrove** [comment]
@alippai A new input is created for each batch and the same accumulator is used across batches, so I think it needs to be this way.

**Comment by andygrove** [comment]
@paddyhoran I have two LGTMs from contributors (not commiters) so if you're OK rubber stamping this I can go ahead and merge and this will allow me to review the other three pending PRs which build on this.

**Review by paddyhoran — APPROVED** [review]

**Pull request closed without merging**
